### PR TITLE
BHBC-1594: Upload Documents / Files to a Project

### PR DESCRIPTION
**Pull request by shreypdev**
# Overview

## Links to jira tickets

- [BHBC-1594](https://quartech.atlassian.net/browse/BHBC-1594)

## This PR contains the following changes

- Authenticated user can upload, view and delete the document.
- Public user can view the document that are uploaded.

## This PR contains the following types of changes

- [x] New feature (change which adds functionality)
- [ ] Enhancement (improvements to existing functionality)
- [ ] Bug fix (change which fixes an issue)
- [ ] Misc cleanup / Refactoring / Documentation
- [ ] Version change

## Checklist

A list of items that are good to consider when making any changes.

_Note: this list is not exhaustive, and not all items are always applicable._

### General

- [ ] I have performed a self-review of my own code

### Code

- [ ] New files/classes/functions have appropriately descriptive names and comment blocks to describe their use/behaviour
- [ ] I have avoided duplicating code when possible, moving re-usable pieces into functions
- [ ] I have avoided hard-coding values where possible and moved any re-usable constants to a constants file
- [ ] My code is as flat as possible (avoids deeply nested if/else blocks, promise chains, etc)
- [ ] My code changes account for null/undefined values and handle errors appropriately
- [ ] My code uses types/interfaces to help describe values/parameters/etc, help ensure type safety, and improve readability

### Style

- [ ] My code follows the established style conventions
- [ ] My code uses native material-ui components/icons/conventions when possible

### Documentation

- [ ] I have commented my code sufficiently, such that an unfamiliar developer could understand my code
- [ ] I have added/updated README's and related documentation, as needed

### Tests

- [ ] I have added/updated unit tests for any code I've added/updated
- [ ] I have added/updated the Postman requests/tests to account for any API endpoints I've added/updated

### Linting/Formatting

- [ ] I have run the linter and fixed any issues, as needed  
       _See the `lint` commands in package.json_
- [ ] I have run the formatter and fixed any issues, as needed  
       _See the `format` commands in package.json_

### SonarCloud

- [ ] I have addressed all SonarCloud Bugs, Vulnerabilities, Security Hotspots, and Code Smells

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes.

## Screenshots

image

image

### CODE REVIEW

File: api/src/services/attachment-service.ts [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+import { GetAttachmentsData } from '../models/project-attachments';
+import { queries } from '../queries/queries';
+import { deleteFileFromS3, generateS3FileKey, getS3SignedURL, uploadFileToS3 } from '../utils/file-utils';
+import { DBService } from './service';
+
+export class AttachmentService extends DBService {
+  async insertProjectAttachment(projectId: number, key: string, file: Express.Multer.File): Promise<{ id: number }> {
+    const sqlStatement = queries.project.postProjectAttachmentSQL(file.originalname, file.size, projectId, key);
+
+    if (!sqlStatement) throw new HTTP400('Failed to build SQL insert statement');
```
Comment (NickPhura) [inline api/src/services/attachment-service.ts]: Can we add the brackets here (and wherever else). Its generally better, as far as I've read, to have them even for 1 liners just to help with legibility and reduce the chance of errors by maintainers in future.

Comment (shreypdev) [reply api/src/services/attachment-service.ts]: Sure thing.

**Review by anissa-agahchen — CHANGES_REQUESTED** [review]
@shreypdev the typical behaviour in SIMS is to overwrite a file is the file name is the same. 
Screen Shot 2022-02-24 at 2 35 47 PM

File: api/src/paths/project/{projectId}/attachments/list.test.ts [hunk 2]
```diff
@@ -38,141 +37,11 @@ describe('lists the project attachments', () => {
     sinon.restore();
   });
 
-  it('should throw a 400 error when no sql statement returned for getProjectAttachmentsSQL', async () => {
```
Comment (anissa-agahchen) [inline api/src/paths/project/{projectId}/attachments/list.test.ts:41]: why remove this test?

File: api/src/paths/project/{projectId}/attachments/upload.ts [hunk 3]
```diff
@@ -142,72 +144,3 @@ export function uploadMedia(): RequestHandler {
     }
   };
 }
-
```
Comment (anissa-agahchen) [inline api/src/paths/project/{projectId}/attachments/upload.ts:145]: is this test not needed anymore?

File: api/src/paths/project/{projectId}/attachments/{attachmentId}/delete.test.ts [hunk 4]
```diff
@@ -83,128 +84,15 @@ describe('deleteAttachment', () => {
     }
   });
 
-  it('should throw a 400 error when no sql statement returned for unsecureAttachmentRecordSQL', async () => {
```
Comment (anissa-agahchen) [inline api/src/paths/project/{projectId}/attachments/{attachmentId}/delete.test.ts:86]: same question here .... not needed?

File: api/src/paths/project/{projectId}/attachments/{attachmentId}/delete.ts [hunk 5]
```diff
@@ -102,39 +95,3 @@ export function deleteAttachment(): RequestHandler {
     }
   };
 }
-
-const unsecureProjectAttachmentRecord = async (securityToken: any, connection: IDBConnection): Promise<void> => {
```
Comment (anissa-agahchen) [inline api/src/paths/project/{projectId}/attachments/{attachmentId}/delete.ts:106]: not needed?

**Review by anissa-agahchen — COMMENTED** [review]
@shreypdev the functionality is all good, just noticed that some tests were removed ... is that intentional?

File: api/src/paths/project/{projectId}/attachments/list.test.ts (hunk 2 above)
Comment (shreypdev) [reply api/src/paths/project/{projectId}/attachments/list.test.ts:41]: This test is moved to project service as no sql error is not thrown in list.test.ts, now it is thrown by attachment-service.

File: api/src/paths/project/{projectId}/attachments/{attachmentId}/delete.test.ts (hunk 4 above)
Comment (shreypdev) [reply api/src/paths/project/{projectId}/attachments/{attachmentId}/delete.test.ts:86]: Tests for unsecure is not needed as we no more switch between secure and unsecure. Also other delete tests are now located in attachment-service.

File: api/src/paths/project/{projectId}/attachments/{attachmentId}/delete.ts (hunk 5 above)
Comment (shreypdev) [reply api/src/paths/project/{projectId}/attachments/{attachmentId}/delete.ts:106]: No use of this according to the requirements

File: api/src/paths/project/{projectId}/attachments/upload.ts (hunk 3 above)
Comment (shreypdev) [reply api/src/paths/project/{projectId}/attachments/upload.ts:145]: Sorry dont exactly know what you are pointing at here

**Review by NickPhura — APPROVED** [review]

**Review by anissa-agahchen — APPROVED** [review]

**Pull request merged**